src/kartograf/atom_mapper.py

```diff
@@ -863,28 +862,78 @@ def suggest_mapping_from_rdmols(
 
         return mapping
 
+    @staticmethod
+    def _split_component_molecules(component: Union[SmallMoleculeComponent, ProteinComponent]) -> list[Chem.Mol]:
+        """
+        Aims at splitting a component into its disconected components based on the
+        connectivity of the molecules that compose it. Useful for mapping multimer components
+        or proteins with structural waters or similarly.
+
+        This returns a list of ``Chem.Mol`` objects with a prop named `Starting_index` indicating the starting
+        index in the original component.
+        """
+        from rdkit.Chem.rdmolops import GetMolFrags
+        rdmol = component.to_rdkit()
+        index_tuples = []
+        fragments = GetMolFrags(rdmol, asMols=True, sanitizeFrags=True, fragsMolAtomMapping=index_tuples)
+        for fragment, index_tuple in zip(fragments, index_tuples):
+            fragment.SetIntProp("Starting_index", index_tuple[0])
+        return fragments
+
+
     def suggest_mappings(
-            self, A: SmallMoleculeComponent, B: SmallMoleculeComponent
-    ) -> Iterator[AtomMapping]:
+            self, A: Union[SmallMoleculeComponent, ProteinComponent], B: Union[SmallMoleculeComponent, ProteinComponent]
+    ) -> Iterator[LigandAtomMapping]:
         """ Mapping generator - Gufe
         return a generator for atom mappings.
 
         Parameters
         ----------
-        A : SmallMoleculeComponent
+        A : Union[SmallMoleculeComponent, ProteinComponent]
             molecule A to be mapped.
-        B : SmallMoleculeComponent
+        B : Union[SmallMoleculeComponent, ProteinComponent]
             molecule B to be mapped.
 
         Returns
         -------
         Iterator[AtomMapping]
             returns an interator of possible atom mappings.
         """
-        yield LigandAtomMapping(
-            A,
-            B,
-            self.suggest_mapping_from_rdmols(
-                molA=A.to_rdkit(), molB=B.to_rdkit()
-            ),
-        )
+        if type(A) != type(B):
+            raise ValueError(f"The components {A} and {B} were not of the same type, please check the inputs.")
+        # 1. identify Component Chains if present
+        component_a_chains = KartografAtomMapper._split_component_molecules(A)
+        component_b_chains = KartografAtomMapper._split_component_molecules(B)
```

Comment:
What happens when the length of these chains is not equal? i.e. should we guard against that case?

Reply:
Yes probably but this would also stop the case when one has more waters (or some other molecule) than the other which should probably still work, but maybe its simpler if we just ensure they are the same length for now?

Reply:
My initial reaction is "in those case I would expect those bits to be mapped as appearing/disappearing".

I recognise that more discussion is probably necessary - should we maybe add the length check for now and have a discussing at a protocol devs meeting to see what folks would like to get from this?

Reply:
The plan is to block having different numbers of components for now and we will come back to this in future.

Reply:
done in [00709d5](https://github.com/OpenFreeEnergy/kartograf/pull/47/commits/00709d5989ddea3005a226e1f8927c5da6780935)